src/glmtools.jl

```diff
@@ -391,3 +420,12 @@ loglik_obs(::Gamma, y, μ, wt, ϕ) = wt*logpdf(Gamma(inv(ϕ), μ*ϕ), y)
 loglik_obs(::InverseGaussian, y, μ, wt, ϕ) = wt*logpdf(InverseGaussian(μ, inv(ϕ)), y)
 loglik_obs(::Normal, y, μ, wt, ϕ) = wt*logpdf(Normal(μ, sqrt(ϕ)), y)
 loglik_obs(::Poisson, y, μ, wt, ϕ) = wt*logpdf(Poisson(μ), y)
+# in principle, μ = θp/(1-p) ==> p = μ/(μ+θ), and thus we should have been able
+# to do following
+# loglik_obs(d::NegativeBinomial, y, μ, wt, ϕ) = wt*logpdf(NegativeBinomial(d.r, μ/(μ+d.r)), y)
+# however, that does not give the correct answer compared to that of R's `glm.nb`
```

Comment:
@nalimilan : I appreciate the encouragement.

W.r.t. the above `loglik_obs` issue, if we use the straightforward 
```
loglik_obs(d::NegativeBinomial, y, μ, wt, ϕ) = wt*logpdf(NegativeBinomial(d.r, μ/(μ+d.r)), y)
```
then the `loglikelihood` numbers are not correct anymore. In `runtest.jl`, the `gm18` model would return
```
julia> loglikelihood(gm18)
-5208.360845881062
```
instead of the expected ` -553.259` (the same LL returned by `glm.nb` from `MASS`). I haven't been able to figure out why.

Reply:
The only other `R` package I found is `vglm`. Its output is similar to that of `glm.nb`:

```
> nb = glm.nb(Days ~ Eth + Sex + Age + Lrn, data=quine)                          
> nb2 = vglm(Days ~ Eth + Sex + Age + Lrn, family=negbinomial(), data=quine)                          
> logLik(nb)
'log Lik.' -546.5755 (df=8)
> logLik(nb2)
[1] -546.5755
```

Reply:
If you want more R-implementation comparisons, there are versions in (at least) the  `glmmTMB` and `mgcv` and  `bbmle` packages ...

Reply:
Many thanks @bbolker . will try them out

Reply:
Report from the `glmmTMP` package: logLik agrees with `glm.nb` and `vglm`
```
> nb_tmb = glmmTMB(Days ~ Eth + Sex + Age + Lrn, family=nbinom2(), data=quine)                          
> summary(nb_tmb)
 Family: nbinom2  ( log )
Formula:          Days ~ Eth + Sex + Age + Lrn
Data: quine

     AIC      BIC   logLik deviance df.resid 
  1109.2   1133.0   -546.6   1093.2      138 


Overdispersion parameter for nbinom2 family (): 1.27 

Conditional model:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  2.89459    0.22793  12.700  < 2e-16 ***
EthN        -0.56938    0.15761  -3.613 0.000303 ***
SexM         0.08231    0.16468   0.500 0.617211    
AgeF1       -0.44843    0.23760  -1.887 0.059119 .  
AgeF2        0.08808    0.24155   0.365 0.715370    
AgeF3        0.35691    0.24662   1.447 0.147840    
LrnSL        0.29211    0.18294   1.597 0.110314    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
```

Reply:
Mmm... It's interesting that R packages don't use the built-in `dnbinom` function. Could it be due to using different parameterizations? To be clear, which [formulation](https://en.wikipedia.org/wiki/Negative_binomial_distribution#Alternative_formulations) do you use in the hand-written formula? Looking at glmmTB, it uses `dbinom_robust` from the [TMB](https://github.com/kaskr/adcomp) package (but you can't read the code yourself since it's GPL2 and GLM.jl is MIT), which is formulated in terms of `mu`. OTC, the parameterization used by Distributions.jl comes from R's `dbinom`, which is formulated in terms of `prob` by default. There's a `dnbinom_mu` variant formulated in terms of `mu` which can be used via an argument in R (see [details in the docs](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/NegBinomial.html)). It can be access from Julia via something like (arguments are `x`, `size`, `mu` and `give_log`, and you'll need to adapt the path to the file):
```
ccall((:dnbinom, ".julia/Rmath/deps/usr/lib/libRmath-julia.so"), Float64, (Float64, Float64, Float64, Int), y, length(y), μ, true)
```

Reply:
The hand-written formula uses the formulation in Section 7.4 of [Modern Applied Statistics with S](http://www.bagualu.net/wordpress/wp-content/uploads/2015/10/Modern_Applied_Statistics_With_S.pdf). I'm not sure if the pdf is legal, just grabbed it off of a Google search.

Reply:
OK. It turns out that at the end of [this section](https://en.wikipedia.org/wiki/Negative_binomial_distribution#Alternative_formulations) of the Wikipedia article, there's an explanation of the relationship between this formula and the standard expression for the negative binomial PDF (but with `m = μ`, `k = y` and `r = θ`). So indeed `p = μ/(μ+θ)` sounds correct.

I've been able to replicate the values of your `loglik_obs` definition in both R and Julia:
```R
> dnbinom(300, mu=100, size=1, log=T)
[1] -7.60022

> dnbinom(300, prob=1/(100+1), size=1, log=T)
[1] -7.60022
```

```julia
julia> function loglik_obs(d::NegativeBinomial, y, μ, wt, ϕ)
           θ = d.r
           wt * (lgamma(θ + y) - lgamma(θ) - lgamma(y + 1) + θ * log(θ) + y * log(μ + (y == 0)) - (θ + y) * log(θ + μ))
       end
loglik_obs (generic function with 2 methods)

julia> loglik_obs(NegativeBinomial(1, 1/(100+1)), 300, 100, 1, 0)
-7.600219772791661

julia> logpdf(NegativeBinomial(1, 1/(100+1)), 300)
-7.600219772791684
```

I guess this should allow replacing the custom method with the standard one?

Reply:
Oh, I know why; it was a stupid mistake of mine. The reparameterization should have been p = θ/(μ+θ), *not* p = μ/(μ+θ). Now we can use `logpdf` and everything looks fine. Apologies for wasting your time on this; thanks much for digging into the details.